packages/flutter_tools/lib/src/version.dart

```diff
@@ -64,110 +64,147 @@ Channel? getChannelForName(String name) {
   return null;
 }
 
-class FlutterVersion {
+abstract class FlutterVersion {
   /// Parses the Flutter version from currently available tags in the local
   /// repo.
-  ///
-  /// Call [fetchTagsAndUpdate] to update the version based on the latest tags
-  /// available upstream.
-  FlutterVersion({
+  factory FlutterVersion({
     SystemClock clock = const SystemClock(),
-    String? workingDirectory,
-    String? frameworkRevision,
-  }) : _clock = clock,
-       _workingDirectory = workingDirectory {
-    _frameworkRevision = frameworkRevision ?? _runGit(
+    required FileSystem fs,
+    required String flutterRoot,
```

Comment:
From what I can tell: 
* `workingDirectory` was renamed to `flutterRoot` and marked as `required`.
* Callers that previously did not supply a value for `workDirectory` now explicitly supply `Config.flutterRoot`, which, ultimately, was used as the default in downstream calls (e.g. `_runGit`).

If I have this correct, would it make sense to also mark `workingDirectory` as `required` in `_runGit` and `determine` (and anywhere else I might be missing)?

Reply:
what would the advantage of making it required be?

Reply:
Sorry to answer with a question, but why was `flutterRoot` made required?

Reply:
Ok, so the one callsite where `flutterRoot` will *not* be `Cache.flutterRoot!` is in `flutter upgrade` or `flutter downgrade`, where you can provide a CLI arg `--working-directory`. So you're right, I should make it required in `_runGit()` and pass it everywhere.